external-crates/move/crates/move-vm-profiler/src/lib.rs

```diff
@@ -240,15 +246,17 @@ macro_rules! profile_open_frame {
 #[macro_export]
 macro_rules! profile_open_frame_impl {
     ($profiler:expr, $frame_name:expr, $gas_rem:expr) => {
-        #[cfg(debug_assertions)]
+        #[cfg(feature = "gas-profiler")]
```

Comment:
Does it make sense to use -- `move_vm_profiler::gas_profiler_feature!` for this (everywhere, in all crates)?  That way, if the set-up is broken and some crates are not correctly propagating the `gas-profiler` feature, it will turn off the feature everywhere, rather than creating a split brain scenario where some crates are behaving as if profiling is enabled, and some are not.

Reply:
If we do get into the split scenario, the consequence will be that the profiler doesn't work, which we would ideally fix asap. Because the features are each additive, it wouldn't  cause build errors, and at worst would be enabling dead code

Reply:
The aim with this suggestion is not to make it possible to keep working in the inconsistent state, it's to make sure that if the set-up is broken in that way, it breaks in a clear and predictable way (and all crates agree that something strange has happened, which is that they asked for the profiler feature to be enabled but it is not) by making the macro you created the single component responsible for deciding whether profiling is enabled or not.

> Because the features are each additive, it wouldn't cause build errors, and at worst would be enabling dead code

Certainly there won't be any build failures because CI will pick them up, but having each crate use its own feature to decide whether the profiling is enabled can cause logical errors that don't manifest in CI -- the affected code won't necessarily be dead.  Imagine the following scenario:

- Crate `a` depends on the profiler crate propagating the `gas-profiler` flag.
- If profiling is enabled, `a::foo` 
  - runs the profiler, and assuming it returns successfully
  - it reads the file from the filesystem, which it can assume succeeds because the profiler run succeeded.

If `a`'s "profiling is enabled" check is based on its own feature flag, and its feature flag propagation was broken, then "runs the profiler" becomes a nop, and reading from the filesystem -- which should succeed -- will fail at runtime.

To protect against this, you need to write an E2E test that covers every scenario that might rely on gas profiling bein enabled.  These are complicated to write and maintain (often they are flaky for CI as well).

If, instead, you used the macro-based approach, you can write a unit/build-time test to make sure that when `a`'s feature flag is enabled, the macro reports success:

```rust
#[cfg(feature = "gas-profiler")]
fn test_profiler_enabled() {
    move_vm_profiler::gas_profiler_disabled! {
        panic!("Didn't successfully enable profiler");
    }
}
```

Reply:
The gas profiler is a structure that stores information upon calls to one of the 4 macros, `profile_close_frame, profile_close_instr, profile_open_frame, profile_open_instr, those are all called from whichever version of the vm's interpreter, so a single package. The profile is only written to file at the end of execution, and we never attempt to read this file. 

I had previously added a test like this in the sui package: 
```
    assert!(!called);
    move_vm_profiler::gas_profiler_feature! {
        called = true;
    }
    assert!(called);
```
I thought about adding tests to the other packages like this but it would mean enabling the profiler as a dev dependency for other packages, and in turn causing us to create profiles willy nilly by all other executed digests in the test suite.


To dig into your concerns a bit deeper, let's think about the feature graph where the sui package is the source of the feature graph, and the skink is the profiler package. The edges there are several intermediate packages with edges between each other, resulting in a graph with several paths from the source to the sink. As long as there remains a path from the source to the sink, the profiler will either work, or throw a build error if the removed edge is from a package that depends on a gated function, struct, etc, and this test is successfully checking the existence of connectivity. 

Your raised concerns here can be modeled as paths removed from the graph. The graph is very strongly connected, and if you remove most any of the edges, which I have tested by going through and deleting combinations of the actual edges, if the source and sink are both connected, the test passes, and the before and after profiles created have no difference from a diff. 

If we remove the edges such that a package is now seemingly no longer traversable starting from the source, but still has a path to the sink, according to my testing the profiler still produces the same profile. Before adding to the hakari exclusion list, if I were to have removed those edges I would get build errors. Now because there are multiple feature-set versions being built, it appears there is another copy of the library being used which has the feature enabled when it is required by a package that depends on it that has the feature enabled. If we try to find these by printing the cargo tree without duplicates, that output is actually insanely long and not very useful to make sense of. But I went through and manually tried to break things, and was pleasantly surprised that it didn't break.

And lastly, if a package gets all edges to it removed, it will result in dead code.